### ISSUE TITLE
keyboard.press_key / keyboard.release_key not working?

### ISSUE THREAD

**Issue by mg979** [autokey/autokey#117]
## Classification:

Bug

## Reproducibility:

Always

## Summary

I can't get keyboard.press_key / keyboard.release_key to work.

## Steps to Reproduce

Script:
```python
keyboard.press_key("<ctrl>")
keyboard.send_key("<home>")
keyboard.release_key("<ctrl>")
```

Keybinding: ctrl+KeyPad[7]

## Expected Results

remap `ctrl+7` to `ctrl+home`

## Actual Results

simple "home" key, at least `keyboard.press_key("<ctrl>")` is ignored

```
2018-01-06 01:28:19,971 DEBUG - iomediator - <ctrl> pressed
2018-01-06 01:28:21,376 DEBUG - service - Raw key: '<np_home>', modifiers: ['<ctrl>'], Key: 7
2018-01-06 01:28:21,376 DEBUG - service - Window visible title: b'~/.themes/sylpheed-dark \xe2\x80\xa2 (FuzzyFileNav) - Sublime Text (UNREGISTERED)', Window class: 'sublime_text.Sublime_text'
2018-01-06 01:28:21,376 DEBUG - interface - using official python-xlib
2018-01-06 01:28:21,376 INFO - service - Matched hotkey phrase/script with prompt=False
2018-01-06 01:28:21,376 DEBUG - service - Script runner executing: Script('home')
2018-01-06 01:28:21,377 DEBUG - service - Ignored locking error in handle_keypress
2018-01-06 01:28:21,378 DEBUG - interface - Send special key: ['<home>']
2018-01-06 01:28:21,385 DEBUG - interface - Recorded keymap change event
2018-01-06 01:28:21,392 DEBUG - iomediator - <ctrl> pressed
2018-01-06 01:28:21,392 DEBUG - iomediator - <ctrl> released
2018-01-06 01:28:21,586 DEBUG - interface - Ungrabbing hotkey: ['<alt>', '<ctrl>', '<shift>'] 'k'
2018-01-06 01:28:21,587 DEBUG - interface - Ungrabbing hotkey: ['<super>'] '<f12>'
2018-01-06 01:28:21,587 DEBUG - interface - Ignored keymap change event
2018-01-06 01:28:21,613 DEBUG - interface - Ungrabbing hotkey: ['<ctrl>'] '<f7>'
```
## Version

0.94
Distro: Manjaro

**Comment by aizigao** [comment]
same problem, has some solution ? 
use manjaro too
I want set 
```
# Enter script code
keyboard.press_key('<shift>')
keyboard.send_key('<insert>')
keyboard.release_key('<shift>')
```

**Comment by josephj11** [comment]
@aizigao Unrelated to this issue, but for your use case, couldn't you just use
`keyboard.send_keys("<shift>+<insert>")`
@mg979 This might also work for you unless your real use case is more complex.
`keyboard.send_keys("<ctrl>+<home>")`

**Comment by josephj11** [comment]
@mg979 Your trace log shows some pretty strange behavior. AutoKey is doing everything you asked it to, but not in the correct order. It's emitting the home keypress first and then pressing and releasing the control key. It looks like a scheduling error of some sort, but I don't know the application source code.

**Comment by josephj11** [comment]
Just to clarify: Are both of you using AutoKey 0.93.x? If not, I'd really encourage you to do so if possible.
It has lots of bug fixes compared to the earlier releases and a new feature or two. It is the only release under development - where the developer will look at problems like this if we can't solve them as users.
Please see our [problem reporting guide](https://github.com/autokey/autokey/wiki/Problem-Reporting-Guide) for what we need to help solve problems quickly.

**Comment by venkatcc** [comment]
I'm seeing the same issue with 0.95.1 as well.

This is my script:
```
keyboard.press_key('<ctrl>')
keyboard.send_keys('gog')
keyboard.release_key('<ctrl>')
```

And the logs generated:

```
2018-09-29 22:42:02,306 DEBUG - iomediator - Key.ALT pressed
2018-09-29 22:42:04,299 DEBUG - service - Raw key: '5', modifiers: [<Key.ALT: '<alt>'>], Key: 5
2018-09-29 22:42:04,299 DEBUG - service - Window visible title: 'Untitled * — Kate', Window class: 'kate.kate'
2018-09-29 22:42:04,300 INFO - service - Matched hotkey phrase/script with prompt=False
2018-09-29 22:42:04,300 DEBUG - service - Script runner executing: Script('GOG')
2018-09-29 22:42:04,301 DEBUG - iomediator - Send via event interface
2018-09-29 22:42:04,302 ERROR - service - Ignored locking error in handle_keypress
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/autokey/service.py", line 206, in __tryReleaseLock
    self.configManager.lock.release()
RuntimeError: release unlocked lock
2018-09-29 22:42:04,303 DEBUG - interface - Recorded keymap change event
2018-09-29 22:42:04,326 DEBUG - interface - Sending string: 'gog'
2018-09-29 22:42:04,328 DEBUG - iomediator - Key.CONTROL pressed
2018-09-29 22:42:04,328 DEBUG - iomediator - Key.CONTROL released
```

This is on Linux Mint 19 Tara.

**Comment by josephj11** [comment]
@venkatcc Just out of curiosity, what does kate do when you hold down Ctrl and press multiple keys such as gog? How is that different than sending `"<ctrl>+g<ctrl>+o<ctrl>+g" `with keyboard.send_keys()?

**Comment by venkatcc** [comment]
@josephj11 That's pretty much what I ended up finally using.
```
keyboard.send_keys("<ctrl>+g")
keyboard.send_keys("<ctrl>+o")
keyboard.send_keys("<ctrl>+g")
```

I actually wanted that to work in a custom CLI application (terminal window), and not Kate.

**Comment by bjohas** [comment]
Hello, 

@josephj11 - responding in this thread, as I'm getting same issue.

## Classification:

Bug

## Reproducibility:

Always

## Version

AutoKey version: 0.95.7

Used GUI (Gtk, Qt, or both): Gtk

If the problem is known to be present in more than one version, please list all of those.

Installed via: deb

Linux Distribution: Ubuntu 19.04

## Summary

keyboard.press_key does not work as expected.

## Steps to Reproduce (if applicable)

I have this script:
```
keyboard.press_key('<ctrl>')
keyboard.press_key('<shift>')
keyboard.send_keys('f');
keyboard.release_key('<shift>')
keyboard.release_key('<ctrl>')
```
## Expected Results
I am expecting the same function as 
```
keyboard.send_keys("<ctrl>+<shift>+f")
```

## Actual Results

keyboard.press_key seems to be ignored. The command types 'f'.

(Note: The reason I need press key is because I need to be able to send two keys - this is for google docs comments navigation, where ctrl/alt need to be held while n and c are pressed in sequence, i.e. the script I need is
```
keyboard.press_key('<ctrl>')
keyboard.press_key('<shift>')
keyboard.send_keys('nc');
keyboard.release_key('<shift>')
keyboard.release_key('<ctrl>')
```
which (in google docs) is not the same as 
```
keyboard.send_keys("<ctrl>+<alt>+n")
keyboard.send_keys("<ctrl>+<alt>+c")
```

**Comment by josephj11** [comment]
@bjohas Just to clarify, I assume you mean `<shift>`, not `<alt>`  in your final example.
This is as close as I can get to what you want
`keyboard.send_keys("<ctrl>+N<ctrl>+C")`  - which will probably not do what you desire.

It does this:

```
2019-08-28 07:26:56,990 DEBUG - service - Raw key: 'o', modifiers: [<Key.SUPER: '<super>'>], Key: o
2019-08-28 07:26:56,990 DEBUG - service - Window visible title: 'Shell — Konsole', Window class: 'konsole.konsole'
2019-08-28 07:26:56,993 INFO - service - Matched Script "Test_syntax" with hotkey and prompt=False
2019-08-28 07:26:56,993 DEBUG - service - Script runner executing: Script('Test_syntax')
2019-08-28 07:26:56,998 ERROR - service - Ignored locking error in handle_keypress
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/autokey/service.py", line 207, in __tryReleaseLock
    self.configManager.lock.release()
RuntimeError: release unlocked lock
2019-08-28 07:26:56,999 DEBUG - iomediator - Send via event interface
2019-08-28 07:26:57,005 DEBUG - interface - Send modified key: modifiers: ['<ctrl>'] key: N
^N2019-08-28 07:26:57,014 DEBUG - interface - Send modified key: modifiers: ['<ctrl>'] key: C
^C2019-08-28 07:26:57,384 DEBUG - iomediator - Key.SUPER released
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/autokey/qtapp.py", line 368, in customEvent
    def customEvent(self, event):
KeyboardInterrupt
```

I don't remember the specifics, but I believe there's something in the source code about those two functions not working together because one of them resets modifiers when it finishes.

Running a couple of different sequences during an AutoKey [trace](https://github.com/autokey/autokey-python2/wiki/Problem-Reporting-Guide) might show the modifiers getting reset.

I can't look into it now, but maybe calling out to `xdotool` can do what you desire.

**Comment by bjohas** [comment]
Thanks for this - unfortunately it's actually &lt;alt>+&lt;ctrl>+CN, rather than &lt;shift>. I'll have a look at the traces and also see whether I can work around with xdotool.

**Comment by Yekutiel** [comment]
Essentially I am experiencing the same problem. When I run the following in a text editor the lowercase letter d appears five times. In other words I see.... `ddddd` Apparently AutoKey is ignoring the `ctrl`.

[API Examples](https://github.com/autokey/autokey/wiki/API-Examples#keyboard)

```
keyboard.press_key('<ctrl>')
keyboard.send_key('d', repeat=5)
keyboard.release_key('<ctrl>')
```

**Comment by bjohas** [comment]
Hi @josephj11 - thanks for adding labels. Would be great to have this looked into.

**Comment by josephj11** [comment]
Thomas does all the coding. He's active and reviews bugs frequently. I'm sure he'll look at this when he can. He could use a helper or two to get to things faster.

**Comment by luziferius** [comment]
The issue with this is simple: `send_keys()` is _synchronous_, while `press_key()` and `release_key()` are _asynchronous_. This means that the "press key" gets enqueued in an event queue for processing later, while send_keys starts to type directly. Because the script is currently the active thread, it simply continues with typing using `send_keys()` after it enqueued `press_key()`.

This is the call chain for  `press_key()`:
https://github.com/autokey/autokey/blob/13d3b1074b524ad7c15d8636f79672794aefac5e/lib/autokey/scripting.py#L125-L134
https://github.com/autokey/autokey/blob/13d3b1074b524ad7c15d8636f79672794aefac5e/lib/autokey/iomediator/_iomediator.py#L188-L190
https://github.com/autokey/autokey/blob/13d3b1074b524ad7c15d8636f79672794aefac5e/lib/autokey/interface.py#L810-L811

Compared to `send_keys()`:
https://github.com/autokey/autokey/blob/13d3b1074b524ad7c15d8636f79672794aefac5e/lib/autokey/scripting.py#L97-L110
https://github.com/autokey/autokey/blob/13d3b1074b524ad7c15d8636f79672794aefac5e/lib/autokey/iomediator/_iomediator.py#L124-L161

The whole chain is a bit longer, but it is completely synchronous.


There are two solutions: Either make `press_key()` and `release_key()` synchronous, or `send_keys()` asynchronous. I’d prefer the first, as that causes less pain for users. If it’s synchronous, you don’t have to worry and introduce artificial delays.

**Comment by Yekutiel** [comment]
Thanks for the explanation. I understand it conceptually but I don't know how to make `press_key() `and `release_key()` synchronous. Therefore, please give me an example of how to make `press_key() `and `release_key()` synchronous.

As I indicated above...

When I run the following in a text editor the lowercase letter d appears five times. In other words I see.... `ddddd` Apparently AutoKey is ignoring the `ctrl`.

[API Examples](https://github.com/autokey/autokey/wiki/API-Examples#keyboard)

```
keyboard.press_key('<ctrl>')
keyboard.send_key('d', repeat=5)
keyboard.release_key('<ctrl>')
```

That last part...

```
keyboard.press_key('<ctrl>')
keyboard.send_key('d', repeat=5)
keyboard.release_key('<ctrl>')
```

is an example. Please provide me with a similar sort of example.

**Comment by luziferius** [comment]
Well, you cant make it synchronous from your side. It needs a code change.
Until this is fixed, a `time.sleep(0.1)` before `send_keys()` should work. 

```python
keyboard.press_key('<ctrl>')
time.sleep(0.1) # may need tweaking
keyboard.send_key('d', repeat=5)
keyboard.release_key('<ctrl>')
```
I hane not tested this, it should pause the script to give the event queue time to empty itself and do everything enqueued.


Having non-working API examples is pretty bad, though... 


The fix in AutoKey is easy, just remove the call to `__enqueue()` in the snipped above. But I don't know why this was done and if removing it causes unintended side effects.

**Comment by bjohas** [comment]
I'm trying to do ctrl+alt+nc. In Google Docs, this requires the ctrl+alt to be held while n and c are typed sequentially. I've used @luziferius' suggestin, but the following sequence doesn't work for me:
```
keyboard.press_key('<ctrl>')
keyboard.press_key('<alt>')
time.sleep(0.5)
keyboard.send_key('n');
time.sleep(0.5)
keyboard.send_key('c');
time.sleep(0.5)
keyboard.release_key('<alt>')
keyboard.release_key('<ctrl>')
```
Any ideas?

**Comment by Yekutiel** [comment]
I tried `alt+d` in a text editor as @luziferius suggested, but it didn't work for me:
```
keyboard.press_key('<alt>')
time.sleep(0.1) # may need tweaking, I also tried 0.5 and 1.0
keyboard.send_key('d', repeat=5)
keyboard.release_key('<alt>')
```

**Comment by hlechner** [comment]
This issue affect me as well,

Im trying to do: (alt) + (left mouse click)

not sure how to do it without press_key

**Comment by josephj11** [comment]
I just filed #391 which is another duplicate of this. I didn't see this one when I looked.

**Comment by kreezxil** [comment]
This is still a problem with AutoKey 0.96 beta 5, a discussion just occurred in the Google Groups support for AutoKey on this very subject. I tested this and this is still happening. I guess that Thomas hasn't gotten around to fixing it. Based on the number of issues related to it, I'm thinking that either most people who want this functionality either semperfy and use send_keys, or they move on to some other way of doing macros and scripts on Linux.

Using what is discussed here, this script fires, but it fires once:
```python
keyboard.press_key('<ctrl>')

for repeat in range(5):
    keyboard.press_key('d')
    keyboard.release_key('d')

keyboard.release_key('<ctrl>')
```

Meaning that asynchronous and synchronous commands can't be mixed in a script, at least not within a code block. My own logs showed that when send_keys was used in place of the loop, that the control down and up happened after the key was sent  5 times.

**Comment by josephj11** [comment]
@kreezxil Thomas ghosted us over a year ago. We don't know why. There were no conflicts... He has been almost inactive on all of GitHub, so maybe something personal happened.

It is a known limitation of AutoKey that synchronous and asynchronous API calls do not cooperate. I think it might have something to do  with independent processes. There's not a single serialized pipeline for events as one might expect.

Do I understand correctly that your final script works as desired? I'm not clear on what you mean by "but it fires once". Do you get 5 Ctrl+D keyboard events?

**Comment by kreezxil** [comment]
Using press_key and release_key and a for loop, the script generated one
CTRL+d, I even added a sleeper, and that didn't help at all. But it leads
to a way to do it. The send_keys method is far superior since it's the one
that works reliably and as noted in the many contributed scripts the
predominate chosen way.

**Comment by josephj11** [comment]
@kreezxil Thanks for the clarification. That's actually quite troubling. With the added delay in the loop, it certainly should have worked. Either the delay has to be excessively long or something really strange is going on in the libraries we use for press and release.

Where did you put the delay? Maybe it needs to be between the press and the release as well as after the release before the next press. I would try one second delays for both because that would presumably be way longer than any race condition that might be occurring.

It would be interesting to see the output of a trace of this to see AutoKey doing the  successive API calls to no effect. There's a small chance that some other messages might appear there with clues as to what is happening.

Someone skilled and tenacious would probably have to sift through an strace to see what was happening beyond the borders of AutoKey. Or maybe an upstream bug needs to be filed somewhere.

My tentative conclusion is that press and release should be in the same category as keyboard.send_key() - something that should only be used when nothing else will work.

This is particularly strange as @Elliria has produced a number of examples where multiple press and releases seem to work as expected. The only obvious difference is that I don't think her examples repeatedly access the same key with no other key presses in between.

**Comment by kreezxil** [comment]
I put it if I recall at the end of the last 'd' in the loop.

We can test this ourselves using CTRL+ESC, there is definitely a huge delay when trying to rapid fire it. Maybe that might be key.

**Comment by josephj11** [comment]
I just tried the following:

```
delay=2.0
keyboard.press_key('<ctrl>')

for repeat in range(5):
    time.sleep(delay)
    keyboard.press_key('<escape>')
    time.sleep(delay)
    keyboard.release_key('<escape>')

time.sleep(delay)
keyboard.release_key('<ctrl>')
```

I got some very weird behavior in terms of timing, but it definitely displayed my system activity pop-up multiple times.

I also got three 
Got BadWindow error while requesting window information
messages in the log. This did not cause AutoKey to throw an exception.

Additionally, I first made the mistake of using `<esc>` instead of `<escape>` which did cause AutoKey to throw errors.
That would have been expected, except AutoKey kept running the script like there was no error.

@BlueDrink9 The weirdness noted in the last two paragraphs where I was doing wrong or very questionable things makes me wonder why AutoKey didn't get more upset and terminate the script. Are these bugs I should file issues about? What was supposed to happen?

**Comment by BlueDrink9** [comment]
No idea

**Comment by Elliria** [comment]
> This is particularly strange as @Elliria has produced a number of examples where multiple press and releases seem to work as expected. The only obvious difference is that I don't think her examples repeatedly access the same key with no other key presses in between.

They do. Here's a set of examples I've shared before that works perfectly for me, turning a lower-case a, b, and c into an upper-case A, B, and C. It presses three keys more than once. You should get ABCABC when running it:

```
keyboard.press_key("<shift>")
keyboard.press_key("a")
keyboard.release_key("a")
keyboard.press_key("b")
keyboard.release_key("b")
keyboard.press_key("c")
keyboard.release_key("c")
keyboard.release_key("<shift>")

# The above does the same thing as this:
keyboard.send_keys("<shift>+a<shift>+b<shift>+c")
```

**Comment by Elliria** [comment]
Oddly enough, this works, but only *after* I click the AutoKey window with the mouse after triggering the script or *after* I roll the mouse wheel without clicking anything after triggering the script or when I trigger the script with no active window:
```
keyboard.fake_keypress("<ctrl>")
keyboard.fake_keypress("<escape>")
```
It's the active window that seems to be interfering with it. I think that AutoKey wants to interact with the active window and only will interact with the operating system if there isn't an active window. Since neither AutoKey nor Geany (which is the text editor I use) have any actions tied to the ctrl+esc hotkey, nothing happens when the script is triggered.

### Some experiments:
_Note that these experiments use the Kate text editor as a neutral program that I just want to open and close to take the focus off of AutoKey. If anyone knows of a simpler way to get rid of all focus, that would make all of these scripts below a heck of a lot prettier._

This works, but only after minimizing AutoKey and then triggering the script:
```
import os
os.system('wmctrl -a "Untitled  — Kate"')
os.system('wmctrl -c "Untitled  — Kate"')
time.sleep(0.5)
keyboard.fake_keypress("<ctrl>")
keyboard.fake_keypress("<escape>")
os.system('wmctrl -a "Untitled  — Kate"')
os.system('wmctrl -c "Untitled  — Kate"')
time.sleep(0.5)
keyboard.fake_keypress("<ctrl>")
keyboard.fake_keypress("<escape>")
```
This works, but only after minimizing AutoKey and then triggering the script:
```
window.activate("kate.kate")
window.close("kate.kate")
time.sleep(0.5)
keyboard.fake_keypress("<ctrl>")
keyboard.fake_keypress("<escape>")
window.activate("kate.kate")
window.close("kate.kate")
time.sleep(0.5)
keyboard.fake_keypress("<ctrl>")
keyboard.fake_keypress("<escape>")
```
This works, but only after minimizing AutoKey and then triggering the script. This one is just prettier than the previous two:
```
def UglyWorkaround():
    window.activate("kate.kate")
    window.close("kate.kate")
    time.sleep(0.5)
    keyboard.fake_keypress("<ctrl>")
    keyboard.fake_keypress("<escape>")
    time.sleep(0.5)

for i in range(4):
    UglyWorkaround()
```
_Note that the last three examples above that all work after taking the focus off of AutoKey will work normally if you trigger them again after giving AutoKey or any other window the active focus. It's the initial run of the script that requires no active window. It's all rather odd._

**Comment by Elliria** [comment]
> Additionally, I first made the mistake of using `<esc>` instead of `<escape>` which did cause AutoKey to throw errors. That would have been expected, except AutoKey kept running the script like there was no error.

That's probably my fault. I'm sorry about that. It looks like I used that as an example in the [Special Keys](https://github.com/autokey/autokey/wiki/Special-Keys) page. I'll fix it right now.

**Comment by andmarti1424** [comment]
Hello. Is this already fixed?
I have the following with unexpected behaviour:
```
keyboard.press_key("<shift>")
keyboard.send_key("<end>")
keyboard.release_key("<shift>")
keyboard.send_key("<delete>")
```
#it just moves to EOL but does not delete any content
thanks

**Comment by kreezxil** [comment]
You can update to the latest version and see if it does, however as we've determined in the discussions, using “keyboard.send_keys” works as intended.

```python
keyboard.send_keys("<shift>+<end>")
keyboard.send_keys("<delete>")
```
And there is a lot less code that way too.

**Comment by andmarti1424** [comment]
@kreezxil Thank you. Works great!

**Comment by josephj11** [comment]
Welcome to the AutoKey community, @andmarti1424 .

Take a look at our wiki for lots of articles and example scripts.
[This one](https://github.com/autokey/autokey/wiki/Key-Combinations) and [this one](https://github.com/autokey/autokey/wiki/Emitting-Keyboard-Events) in particular will explain part of why what you tried didn't work.

Also, some of our API calls just don't mix with each other. `keyboard.press_key()` and `keyboard.release_key()` work fine with each other, but `keyboard.send_keys()` resets them when it starts.

**Comment by josephj11** [comment]
Unless and until we replace our keyboard module with something new for Wayland, I think we can close this as a won't fix.

I'm adding that tag now, but I'll wait for comments before closing it.

@sebastiansam55 @BlueDrink9 @Elliria

**Comment by kreezxil** [comment]
@josephj11 can you turn it into a discussion so that users will stop opening new reports in regard to it?

**Comment by josephj11** [comment]
@kreezxil  That's a good idea. If no one objects, I'll do that. If I just close it, most people would probably never see it. That way, at least some will.

I'm quite sure it won't stop the questions though. :)

**Comment by kreezxil** [comment]
questions are good.

**Comment by Elliria** [comment]
Just as a point of interest, the reason the original code fell over was because the `keyboard.press_key()` API call repeatedly presses the key rather than holding it down steadily. This can be seen if you use it with a letter:
```
keyboard.press_key("a")
time.sleep(2)
keyboard.release_key("a")
```
So, instead of the **Shift** key being pressed once and held down, it produced presses of the **Shift** key and a press of the **End** key, which didn't satisfy the **Shift+End** key-combination.

**Comment by Elliria** [comment]
> Unless and until we replace our keyboard module with something new for Wayland, I think we can close this as a won't fix.

I think that's a good idea. It seems like it isn't actually an AutoKey issue, but is an operating-system issue and is dependent on whether you have **keyboard repeat** enabled on your system or not.

*In Kubuntu, for example, you can go into **K menu** --> **Settings** --> **System Settings** --> **Input Devices** --> **Keyboard** and choose whether to **Repeat the key** or **Do nothing** when a key is held down.*

If you've got **keyboard repeat** enabled, then the `keyboard.press_key()` API call won't work for holding down a key, but not because AutoKey fails to press it.

**Comment by josephj11** [comment]
> the `keyboard.press_key()` API call repeatedly presses the key rather than holding it down steadily

I would only believe that if I saw it in an AutoKey trace. (The trace included above is the result of doing things that don't work in AutoKey, so it isn't evidence of anything.)

What is probably happening is that a keyboard repeat delay and speed is set in the desktop environment (DTE) and holding down a key long enough triggers that behavior.

You can see if that's the case simply by physically holding down a key for a while.

If that's the problem, changing the settings in the DTE should help.

Also, since Shift is almost always just a modifier key, by itself, it shouldn't send anything from the keyboard although its scancode might be sent once. I don't know how things that low level work.

I reread your comment after that one and you seem to agree with me, so eitiher I'm missing something or your two posts contradict each other.

**Comment by Elliria** [comment]
I'd say the second one builds upon the first one rather than contradicting it. In the first one, I described what seemed to be happening and in the second one, I mentioned what I figure is the most likely explanation for why it was happening. If there had been no repeat, there would have been no known reason for that code not to work. The solution that @kreezxil came up with is especially handy because it neatly gets around the issue without requiring the user to have to make system or desktop changes that would cause inconvenience elsewhere (most of us want our keyboards and mice to be set "just so without any changes, thank you very much" and wouldn't want to be messing with them).

**Comment by josephj11** [comment]
Those input settings do come pretty close to dogma. ;)

**Comment by kreezxil** [comment]
knowing that press key spam presses the key, tells me that we know what to look for to fix that. are we dependent on one person to make patches?

**Comment by josephj11** [comment]
@kreezxil Not sure what you mean by "press key spam". Keyboard autorepeat is a DTE feature, not a bug and is an easily modifiable setting once someone knows enough to look for it.

As far as patches/PRs to AutoKey, anyone can submit them and we have occasionally received them from people who are not regular partcipants in the project. To make that easier, we have been revising and improving our contributor documentation.

Sam and Elliria are our only currently active developers. Sam is doing some very important work on several fronts, one of which may lead to AutoKey working under Wayland and  Elliria is just getting started with the AutoKey development process although she seems to be getting more proficient quickly.

We have recently added AutoKey to the Code Triage website and have already seen a couple of people from there show interest in our project. We are hopeful more will do so.

As far as merging PRs - which is what actually makes them part of AutoKey, Sam and I have that authority and Elliria can get it whenever she decides she wants it. Sam knows what he's doing and I have the ability (and the ability to grant it to others) so that we can't get locked out of our own repos. I only run merges when no one else more qualified is available to do so.

**Comment by kreezxil** [comment]
If **keyboard.press_key** is supposed to be a spamming operation, then I think its wiki entry and API entry are in error. Because the wiki explicitly says, the key is held down until the release_key method is called. https://github.com/autokey/autokey/wiki/API-Examples#press-key

but you repeatedly state to those mentioning it doesn't do that but instead spam "that it is supposed to spam the key"

and in the API documentation itself

https://autokey.github.io/api/keyboard.html#autokey.scripting.Keyboard.press_key

That is why I asked what I did because I have half a mind right now to rectify this and push out a fix.

There are use case scenarios where one might want to programmatically press and hold keys for a length of indeterminate time that send_key and send_keys would not be adequate for.

However, if changing the documentation and the wiki are enough to make everything smooth across the board, then I technically don't need to do anything.

**Comment by Elliria** [comment]
> If **keyboard.press_key** is supposed to be a spamming operation, then I think its wiki entry and API entry are in error. Because the wiki explicitly says, the key is held down until the release_key method is called. https://github.com/autokey/autokey/wiki/API-Examples#press-key

I can fix that for now, if you'd like, by mentioning that its behavior is dependent on how the keyboard behavior is configured in your distribution.
 
> That is why I asked what I did because I have half a mind right now to rectify this and push out a fix.

If you know how to fix it so that it can override the distribution's settings if the user has the key-repeat enabled, @kreezxil, that would be wonderful, especially if you can make that an optional setting so that users who do want the spamming behavior (in, for example, a gaming session) can still opt for that.

**Comment by josephj11** [comment]
What seems unclear in this discussion (at least to me) is that the so called "spamming" behavior is a feature of most DTEs and not caused by AutoKey or its dependencies. This is such a basic feature that I believe I remember keyboards that actually had it implemented in hardware/firmware in keyboards themselves. It has always been a user modifiable behavior.

However, since this functionality is in the DTE and there are multiple DTEs, any programmed solution would have to be DTE specific.

**Comment by kreezxil** [comment]
The reason it comes up is those of us that :vomiting_face: use Windows, and AutoHotKey, know that regardless of our keyboards have the other experience. You can emit something that holds the key down, presses other keys to be used with the held down key, and then release that first key. And there is no "quote key spamming endquote".

**Comment by josephj11** [comment]
@kreezxil I assume that means that Windows knows the difference between someone holding a physical key down and a program generating a key down event. It would probably be nice if Linux had that too, but I don't think it does.

**Comment by Elliria** [comment]
I'm going to confuse matters even more. If you run `xev` and then press and hold the **Shift** key, it shows only one press and release. If you press and hold the **a** key, it shows a repeated spam of the key. It's possible that the operating system is intelligent in deciding which keys to spam and which to hold.

Then again, it's also possible that this is a classic example of [Heisenbug](https://en.wikipedia.org/wiki/Heisenbug) behavior that makes it impossible to properly diagnose. Perhaps the environment created by `xev` is conducive to one type of behavior and the environment created by AutoKey is conducive to another type of behavior.

I still suspect that when AutoKey presses the key, it obeys the setting in the operating system regardless of which key is pressed. That would, at least, be one possible explanation for why the key-press that hasn't been released yet fails to act as though it's still pressed.

**Comment by josephj11** [comment]
I don't think AutoKey knows anything about keyboard repeat. Either the DTE does it or it doesn't based on DTE settings. As far as AutoKey is concerned, it's pressing and holding the key as expected, but the DTE has other ideas which override that. It appears that the DTE doesn't distinguish between a key manually pressed by the user and one pressed and held by AutoKey, so it does what it would do when the key is manually pressed - which includes auto repeat if that hasn't been disabled in the DTE.

**Comment by Elliria** [comment]
I agree. If @kreezxil knows of a way to get AutoKey to step in and insist on the key being held down steadily rather than repeated, though, that would make a nice additional option for the times when that's exactly what we would want it to do.

**Comment by josephj11** [comment]
If there is a way to do that, it would be a big deal - beyond the scope of AutoKey as well. Maybe there could be a way to temporarily disable the repeat rather than change it for everything from now on.

**Comment by Elliria** [comment]
No idea, but the alternative way of accomplishing the key-press above works, so it's a way around it. And meanwhile, it sounds like @kreezxil has an idea for how to code it.

**Comment by kreezxil** [comment]
Oh, I don't have an implementation in mind. I'm just feeling extremely motivated based on the discussion and how often the issue comes up elsewhere to go for it.

However, it feels rightly odd to me that 
```python
keyboard.send_keys("<shift>+<end>")
``` 
works as visually apparent vs. what happens when you do

```python
keyboard.press_key("<shift>")
keyboard.press_key("<end>")
keyboard.release_key("<end>")
keyboard.release_key("<shift>")
```

Going through the examples in the wiki, and this discussion thread, I'm going to have to side with @Elliria on this one. This is very much a Heisenberg scenario.

I thought I was beginning to see a pattern, then it faded from view and was once again occluded by the various implementations of the Linux Desktops.

But then again **send_keys** always works and this press and release key is the variable one that doesn't work across all the desktops.

My virtual finger says there's a flag that signifies something as being held down, that the flag is prematurely releasing.

Without looking at the code, how is press_key implemented? does it move commands to a stack and then pop them off when a release is encountered? While stack like in behavior this model wouldn't fit here because the when the user issues a release key out of logical sequence that would cause erratic behavior that should crash the script. Oh, does that mean that press and release key are being trapped? If so, maybe we should unwrap their try/catch traps and let them crash, it would go a long way in debugging user scripts and maybe just review the uncertainty from this thing.

Logically, I would think when press key is encountered the system is told to hold that key down, and then except wherein autokey itself would crash, it would not release that key until it was told to do so. so at best there should be a list where keys are placed that are supposed to be held down, which is where press key would put them. and then when release key is found, we should check that update that list to remove the key being released, and release the key, whether or not the key is in the list would be irrelevant.

However, I can see another issue in this. What if some other script or method on the Linux Environment were also trying to  hold down or release a key?

Seems like the solution would be turn what's in send_keys, it's routines that is, ... break them down into the new press and release key methods. Have we tried anything like that?

I can think of quite a few experiments to try out.

**Comment by josephj11** [comment]
The code is a bit over my head, but it looks like press and release keys are asynchronous API calls - are enqueued somewhere outside of AutoKey. I don't think AutoKey keeps track of them once it does that. I'm not sure what releases a key if AutoKey doesn't explictly do so. But adding delays after each press/release would be the first thing to try.

Right now, Sam is working on replacement code for that whole mechanism with something that may work with Wayland, so any major experimentation/efforts should be coordinated with him. We don't want to waste effort on code that may be going away.

**Comment by Elliria** [comment]
> Oh, I don't have an implementation in mind. I'm just feeling extremely motivated based on the discussion and how often the issue comes up elsewhere to go for it.

Ah, well. If @sebastiansam55 is already working on that whole section, this will probably stop being an issue.
 
> ```python
> keyboard.press_key("<shift>")
> keyboard.press_key("<end>")
> keyboard.release_key("<end>")
> keyboard.release_key("<shift>")
> ```

OMG, this was insane. I ran that code on my production machine and at first I thought it didn't do anything, because nothing happened. Then I pressed my right arrow key a couple of times and it began selecting characters. The problem was that it never stopped. I changed to a different window and it selected all its text. I used the tray icon to close AutoKey. I then opened a text file with some text in it and all of its text was selected. I then went into a terminal window and typed `pkill autokey` to try to end its process, but nothing happened when I hit the **Enter** key (it was probably too busy trying to select things). I finally gave up and used the K menu to reboot the computer.

From that strange romp, I can only guess that the keys got pressed, but never released, so my computer just became selection-happy.

What's odd about this is that I've used the **press_key** and **release_key** API calls successfully in the past. It may well matter which keys are being pressed. Perhaps it behaves well with certain ones, but not others.
 
> Going through the examples in the wiki, and this discussion thread, I'm going to have to side with @Elliria on this one. This is very much a Heisenberg scenario.

Yeah, and those are notoriously hard to figure out and/or fix.
 
> But then again **send_keys** always works and this press and release key is the variable one that doesn't work across all the desktops.

Another one that works really solidly is **fake_keypress**.

Interestingly, **press_key** uses **fake_keydown** and **release_key** uses **fake_keyup**, while **fake_keypress** uses **fake_keypress** and all of them get enqueued in the [interface.py](https://github.com/autokey/autokey/blob/develop/lib/autokey/interface.py) file. My brain is getting tangled up.
 
> My virtual finger says there's a flag that signifies something as being held down, that the flag is prematurely releasing.

Or possibly that it doesn't let up when told to, which triggers some sort of error state that's causing misbehavior. Or, worse yet, both - depending on circumstance.

**Comment by josephj11** [comment]
I'm thinking that it may be some sort of race condition because things are enqueued and run in separate threads.

I would take an offending piece of code and add 1 second delays between each API call and see if that makes a difference.

Sometimes, it even helps to use insanely long delays (of several seconds each). Then you can visually see each step occurring and get a feel for where it goes off the rails.

In this case, maybe another script such as bash could be written to run at the same time that would retrieve the state of the keys invoved and tell you when they change. Not sure how to write that, but xdotool might help.

**Comment by kreezxil** [comment]
Furthermore, now that [Espanso](https://espanso.org) is a thing, we have a way to get a type of other tests. Using its weird syntax could do something with the press and release key, there's other macro/script programs for Linux too. We could make test case scenarios on each. And that way we can determine if the solution is AutoKey intrinsic, or DE intrinsic.

Considering that @Elliria pointed out the interface.py, that seems like a good place to start, and @josephj11 pointed out enqueing and parallel threads. Whichever module controls that would be another place to look. I feel like wherever the enqueing is happening would be the best place to really look. But both places deserve a good brain-bending look.

**Comment by josephj11** [comment]
@kreezxil The problem is that Sam is probably the only one of us who knows enough to start doing that and I _really_ don't want to distract him from what he's currently working on.

**Comment by Elliria** [comment]
Another possibility is that it's properly pressing and releasing the key **once**, so no error is raised, but then unexpectedly pressing the key one more time and never releasing it.

The **fake_keyup** and **fake_keydown** show up in [interface.py](https://github.com/autokey/autokey/blob/develop/lib/autokey/interface.py) and [iomediator.py](https://github.com/autokey/autokey/blob/develop/lib/autokey/iomediator/iomediator.py). We could put a "howdy" into the log in each of the functions that run those to watch them in action. Both **press_key** and **release_Key** are all over the place. The same could be done for them, but it would be an exercise in insanity to chase them all down.

**Comment by josephj11** [comment]
@Elliria That's a argument for having logging levels. What we have now could be level 1 and something like the above could be level 2 ... So it would be there, but do nothing unless specifically enabled in rare situations. The level could be an optional parameter to maintain compatibility with current usage.

**Comment by kreezxil** [comment]
yes, i'm in favor of logging levels. all functions that don't have logging
need to have it so that we can ask for a user to turn it on, and then we
can get better data. Wow, it's the best idea yet in this discussion. ;)

**Comment by josephj11** [comment]
I brought it up before but it didn't get any traction. IDK what else we'd want to log. Someone would have to comb through the code and find anything that would actually be useful to see. Then, an enhancement issue could be opened.

There's currently a logger function in the code, so that would just have to be wrapped in a level test function which could be added anywhere interesting. The only other piece would be setting the level  which could be  a JSON entry without any GUI. So the actual coding sounds pretty trivial. The hard part is finding what to log and where to place the logger calls.

**Comment by kreezxil** [comment]
Not hard, just tedious. Shall we draw straws then? as to who will update their fork, then proceed to add logger calls to all the routines. <--- that's what I'll do. Why pick and choose, just make sure that if enabled it all can be logged. And like you said, levels, so level 1 would get just the main big routines, and level 4 would get everything, assuming we can determine two more levels in there.

**Comment by josephj11** [comment]
We should probably make a feature branch that we can all add to.

The problem is that the new messages have to be helpful and not just spam. While a few might be useful, I can't think of any that I really wished I had although I seem to remember wanting one or two in the past.

More isn't necessarily better. If you've ever looked at the output of strace...

**Comment by Elliria** [comment]
> This is particularly strange as @Elliria has produced a number of examples where multiple press and releases seem to work as expected. The only obvious difference is that I don't think her examples repeatedly access the same key with no other key presses in between.

I've done both with success many times. It's just that sometimes we get the odd behavior we're seeing above.

This code, with no key-presses between the press and release:
```python
keyboard.press_key("e")
time.sleep(1)
keyboard.release_key("e")

```
produces this result:
```
eeeeeeeeeeeeeee
```
This code, with no key-presses between the press and release:
```python
keyboard.press_key("e")
keyboard.release_key("e")
```
produces this result:
```
e
```
This code, with several key-presses between the presses and releases:
```python
keyboard.press_key("<shift>")
keyboard.press_key("a")
keyboard.press_key("b")
keyboard.press_key("c")
keyboard.release_key("a")
keyboard.release_key("b")
keyboard.release_key("c")
keyboard.release_key("<shift>")
```
produces this result:
```
ABC
```
But **this** code, from above:
```python
keyboard.press_key("<shift>")
keyboard.press_key("<end>")
keyboard.release_key("<end>")
keyboard.release_key("<shift>")
```
Temorarily bricked my computer until I managed to wrestle myself free by rebooting the computer to regain control.

I think the misbehavior is in combinations of non-printing keys, like **Shift**, **End**, **Ctrl**, etc.

**Comment by Elliria** [comment]
> If there is a way to do that, it would be a big deal - beyond the scope of AutoKey as well. Maybe there could be a way to temporarily disable the repeat rather than change it for everything from now on.

Nice. [Stack Overflow](https://askubuntu.com/questions/576421/disable-keyboard-repeat-from-command-line) offers a way to do it, which means we could script it before key-presses that we don't want to have repeats of:
* Disable it:
   ```
   xset r off
   ```
* Enable it again with the same settings as before:
   ```
   xset r on
   ```
There are other ways to do it listed on that page, but this should be easy to put into an AutoKey script. Then again, I can't seem to resist breaking things...

This is part of the way there for me, but it only works properly once and then falls over if you keep running it, giving very strange behavior, like not running all the way until you click into the active window with the mouse or not printing enough characters for the two times when there should be repeats, etc. Very strange:
```python
import subprocess 

# Press, delay, and release a key:
keyboard.press_key("a")
time.sleep(1)
keyboard.release_key("a")

# Disable keyboard-repeat:
#subprocess.run(["xset", "r", "off"])
subprocess.Popen(["xset", "r", "off"])

# Press, delay, and release a key:
keyboard.press_key("b")
time.sleep(1)
keyboard.release_key("b")

# Enable keyboard-repeat with the same settings as before:
#subprocess.run(["xset", "r", "on"])
subprocess.Popen(["xset", "r", "on"])

# Press, delay, and release a key:
keyboard.press_key("c")
time.sleep(1)
keyboard.release_key("c")
```
If we can get this to work reliably, then it would put the control into our hands, which would be perfect.

**Comment by Elliria** [comment]
> @Elliria That's a argument for having logging levels. What we have now could be level 1 and something like the above could be level 2 ... So it would be there, but do nothing unless specifically enabled in rare situations. The level could be an optional parameter to maintain compatibility with current usage.

Interesting. I had never heard of those before, but yes, I could see where those would be useful.

**Comment by Elliria** [comment]
> yes, i'm in favor of logging levels. all functions that don't have logging need to have it so that we can ask for a user to turn it on, and then we can get better data. Wow, it's the best idea yet in this discussion. ;)

Yes, I totally agree. I love the idea of being able to turn it on and off. We could probably just add another (or several other) command-line switch(es) for logging levels.

**Comment by josephj11** [comment]
I hate to tell you this because I probably won't hear from you for a week after I do. ;)

But ...

```
ESC [ ? 8 h
              DECARM (default on): Set keyboard autorepeat on.
```

Same thing with an l (lower case L) instead of an h should turn it off.

Here's the rabbit hole:

https://man7.org/linux/man-pages/man4/console_codes.4.html

Not really sure how much of this stuff is guaranteed to work and where - might only be honored by (some) terminal programs.

**Comment by kreezxil** [comment]
> We should probably make a feature branch that we can all add to.
> 
> The problem is that the new messages have to be helpful and not just spam. While a few might be useful, I can't think of any that I really wished I had although I seem to remember wanting one or two in the past.
> 
> More isn't necessarily better. If you've ever looked at the output of strace...

The idea is that one would turn on varying levels of logging to get more and more data, level 0 would be the default producing the stack traces it does now.

**Comment by josephj11** [comment]
@kreezxil I understand the idea. What I need to be convinced of is the utility of additiional trace data. What _specfically_ do you want to see that's not there now and how would that help with an actual problem.

I'm all for things that help, but I don't want to add anything that ends up just being bloat.

**Comment by Elliria** [comment]
No worries, @josephj11. I'm completely lost and you may have written all of that above in a language I don't understand, so it all glided right past me without leaving a mark. Are you saying that pressing the **Escape** key followed by the left square bracket followed by the question mark followed by the **8** key followed by the **h** key would toggle autorepeat?

Also, in this particular scenario, if Sebastian weren't already working on replacing all of this, what we could do would be to find every function that's supposed to be involved with key-presses and key-releases and put a log comment in that says "foo ran successfully" or something like that, with "foo" being the name of the function. Then you run AutoKey and do some key-presses and key-releases. Then look at the log and see if all of the functions ran, if they ran in the right order, and if they ran within the time-frame you would expect them to.

I've done this sort of thing in programs that were giving me a headache and it helped me to figure out that [insert function here] wasn't running, which I hadn't figured out by my other debugging methods.

**Comment by josephj11** [comment]
What I am saying is that DEC was the envy of minicomputers back in the day and their  VT100 terminal became a defacto standard for terminal software to emulate. UNIX was first written on a DEC PDP-8 computer and it probably had a VT100 terminal as its console. It supported lots of escape sequences, one of which turns on keyboard autorepeat and another one that turns it off. The problem being that such things probably only work in terminal emulators not in other software so konsole and xterm should work, but kate or Libreoffice are a whole lot less likely to honor such escape sequences.

```
ESC [ ? 8 h
              DECARM (default on): Set keyboard autorepeat on.

ESC [ ? 8 l
              Set keyboard autorepeat off.
```

As an aside, this lets you do cool stuff in bash like change the color of text written to the terminal. I have a set of scripts that backup files and directories  to Dropbox. When they make a backup, they write a green message and when the backup is already current, they write a red message. When I run these every day, I can tell at  a glance what happened just by looking at the colors of the messages. I'm sharing one with you [here](https://www.dropbox.com/s/2qh8gexvz86chfw/put_pass?dl=0). You can ignore most of the script. Just look at the functions at the beginning of it. They do the magic.

**Comment by Elliria** [comment]
Ah, yes, that's why that flew right over my head. Your `tput` color functions s are very cool. Here's [my `tput` snippet](https://bpa.st/GKVD2) (on the paste site that the IRC Python channel uses) that demonstrates doing them manually one by one.
